### ISSUE TITLE
Tag?

### ISSUE THREAD

**Issue by ChrisRackauckas** [FluxML/Flux.jl#1924]
I'm seeing that the bound Flux has on ArrayInterface is causing a lot of packaging issues.

```julia
(@v1.7) pkg> add Flux@0.12.9 ArrayInterface@5
   Resolving package versions...
ERROR: Unsatisfiable requirements detected for package Flux [587475ba]:
 Flux [587475ba] log:
 ├─possible versions are: 0.4.1-0.12.9 or uninstalled
 ├─restricted to versions 0.12.9 by an explicit requirement, leaving only versions 0.12.9
 └─restricted by compatibility requirements with ArrayInterface [4fba245c] to versions: 0.4.1-0.12.4 or uninstalled — no versions left
   └─ArrayInterface [4fba245c] log:
     ├─possible versions are: 0.0.1-5.0.5 or uninstalled
     ├─restricted to versions 2.7.0-5 by OrdinaryDiffEq [1dea7af3], leaving only versions 2.7.0-5.0.5
     │ └─OrdinaryDiffEq [1dea7af3] log:
     │   ├─possible versions are: 6.7.1 or uninstalled
     │   └─OrdinaryDiffEq [1dea7af3] is fixed to version 6.7.1
     ├─restricted to versions 5 by an explicit requirement, leaving only versions 5.0.0-5.0.5
     └─restricted by compatibility requirements with LoopVectorization [bdcacae8] to versions: [3.1.4-3.2.2, 5.0.1-5.0.5], leaving only versions: 5.0.1-5.0.5
       └─LoopVectorization [bdcacae8] log:
         ├─possible versions are: 0.1.0-0.12.103 or uninstalled
         └─restricted to versions 0.12 by OrdinaryDiffEq [1dea7af3], leaving only versions 0.12.0-0.12.103
           └─OrdinaryDiffEq [1dea7af3] log: see above
```

But it's fixed on master. Could Flux get a tag?

**Comment by CarloLucibello** [comment]
We are one PR away (https://github.com/FluxML/Flux.jl/pull/1921) from tagging v0.13

**Comment by mcabbott** [comment]
We should also consider just removing this dependence. It seems to be loaded for exactly one line, to replace Zygote.Fill with something mutable, which could surely be done in a few lines:

https://github.com/FluxML/Flux.jl/blob/de76e0853db8f0216addaaa4eacf6de39630d834/src/optimise/train.jl#L22-L26

Mutating gradients isn't safe anyway, although perhaps nobody hits the edge cases. Moving to Optimisers.jl should avoid both problems.

**Comment by ChrisRackauckas** [comment]
No that isn't just about mutability. That would also reshape for matrices, apply type information like ComponentArrays, etc.

**Comment by AriMKatz** [comment]
Also if we eventually want to have first class support for  labeled arrays, we'd need ArrayInterface to do it right IIUC.

**Comment by ChrisRackauckas** [comment]
You can think of `restructure` as an expanded idea of reshape, to also include structural information. So for example if `x̄` is a vector and `x` is a ComponentArray of matrices, it would reshape and add back that lost type information. I don't think this specific case is needed for LabelledArrays, but it would use the type of `x` to convert `x̄` to match that as `x̄r` before applying.

**Comment by mcabbott** [comment]
I confess to being amazed that there are tests involving ComponentArrays, this is good. 

But quite what's needed here is not so obvious. Flux wants the gradient to be mutable, but this isn't safe anyway, as mutation may affect another gradient. ChainRules demands that any valid gradient be something you can add to the primal, which is (almost) what Optimisers does, no mutation. Under that scheme, either fancy arrays produce valid gradients, or they are already broken, so hopefully no special handling is needed.

**Comment by cossio** [comment]
Bump (https://github.com/FluxML/Flux.jl/pull/1921 is merged)

**Comment by CarloLucibello** [comment]
we released v0.12.10 and v0.13.0 both supporting ArrayInterface@5
